linked_list/linked_list.py

```diff
@@ -86,32 +174,88 @@ def visit(self):
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def reverse(self):
-        pass
-  
+        # iterative method
+        if self.head is None:
+            return None
+
+        current = self.head
+        previous = None
+
+        while current is not None:
+            saved_next = current.next
+            current.next = previous
+            previous = current
+            current = saved_next
+        self.head = previous
+
     ## Advanced/ Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_middle_value(self):
-        pass
+        length = 0
+        current = self.head
+
+        while current is not None:
+            length += 1
+            current = current.next
+        
+        mid = length //2
+        current = self.head
+        idx = 0
+
+        while idx < mid:
+            idx += 1
+            current = current.next
+        return current.value
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_nth_from_end(self, n):

```

Comment:
👍  Again this works, but we could use some arithmetic and some prior methods like `length` and `get_at_index`.